EIPS/eip-2315.md

```diff
@@ -42,69 +64,91 @@ This is the smallest possible change that provides native subroutines without br
 
 These changes do not affect the semantics of existing EVM code.
 
+## Altenative variants
+
+One possible variant, would be to add an extra clause to the `BEGINSUB` opcode. 
+
+- A `BEGINSUB` opcode may only be reached via a `JUMPSUB`. 
+
+This would make `walking` into a subroutine an error. The rationale for this would be to possibly improve static analysis, being able 
+to make stronger assertions about the code flow. 
+
+This is not part of the current specification, since code-generators can trivially implement these guarantees by always prepending `STOP` opcode before
+any `BEGINSUB` operation. 
+
 ## Test Cases
+
+### Simple routine
+
+This should jump into a subroutine, back out and stop.
+
+Bytecode: `0x6004b300b5b7`
+
+
+|  Pc   |      Op     | Cost |   Stack   |   RStack  |
+|-------|-------------|------|-----------|-----------|
+|    0  |      PUSH1  |    3 |        [] |        [] |
+|    2  |    JUMPSUB  |    8 |       [4] |        [] |
+|    4  |   BEGINSUB  |    1 |        [] |      [ 2] |
+|    5  |  RETURNSUB  |    2 |        [] |      [ 2] |
+|    3  |       STOP  |    0 |        [] |        [] |
+
+### Two levels of subroutines
+
+This should execute fine, going into one two depths of subroutines
+Bytecode: `0x6800000000000000000cb300b56011b3b7b5b7`
+
+|  Pc   |      Op     | Cost |   Stack   |   RStack  |
+|-------|-------------|------|-----------|-----------|
+|    0  |      PUSH9  |    3 |        [] |        [] |
+|   10  |    JUMPSUB  |    8 |      [12] |        [] |
+|   12  |   BEGINSUB  |    1 |        [] |      [10] |
```

Comment:
and `[11]` here?

Reply:
Yes. These were all generated by the actual geth evm, so I'll regenerate them when I update the code